ci/scan_flattened_deps.py

```diff
@@ -0,0 +1,343 @@
+
+  # Extract commit hash, save in dictionary
+  for line in lines:
+    os.chdir(CHECKOUT_ROOT + clone_test_dir)
+    dep = line.strip().split(
+        '@'
+    )  # separate fully qualified dep into name + pinned hash
+
+    common_commit = get_common_ancestor_commit(dep)
+    if isinstance(common_commit, str):
+      queries.append({'commit': common_commit})
+    else:
+      failed_deps.append(dep[0])
+
+  print(
+      'Dependencies that could not be parsed for ancestor commits: ' +
+      ', '.join(failed_deps)
+  )
+  print(
+      'Dependencies that have not been rolled in at least 1 year: ' +
+      ', '.join(old_deps)
+  )
+
+  # Query OSV API using common ancestor commit for each dep
+  # return any vulnerabilities found
+  data = json.dumps({'queries': queries}).encode('utf-8')
+  req = request.Request(osv_url, data, headers=headers)
+  with request.urlopen(req) as resp:
+    res_body = resp.read()
+    results_json = json.loads(res_body.decode('utf-8'))
+    if resp.status != 200:
+      print('Request error')
+    elif results_json['results'] == [{}]:
+      print('Found no vulnerabilities')
+    else:
+      results = results_json['results']
+      filtered_results = list(filter(lambda vuln: vuln != {}, results))
+      if len(filtered_results) > 0:
+        print(
+            'Found vulnerability on {vuln_count} dependenc(y/ies), adding to report'
+            .format(vuln_count=str(len(filtered_results)))
+        )
+        print(*filtered_results)
+        return filtered_results
+      print('Found no vulnerabilities')
+  return {}
+
+
+def get_common_ancestor_commit(dep):
+  """
+  Given an input of a mirrored dep,
+  compare to the mapping of deps to their upstream
+  in DEPS and find a common ancestor
+  commit SHA value.
+
+  This is done by first cloning the mirrored dep,
+  then a branch which tracks the upstream.
+  From there,  git merge-base operates using the HEAD
+  commit SHA of the upstream branch and the pinned
+  SHA value of the mirrored branch
+  """
+  # dep[0] contains the mirror repo
+  # dep[1] contains the mirror's pinned SHA
+  # upstream is the origin repo
+  dep_name = dep[0].split('/')[-1].split('.')[0]
+  with open(DEPS, 'r', encoding='utf-8') as file:
+    local_scope = {}
+    global_scope = {'Var': lambda x: x}  # dummy lambda
+    # Read the content.
+    with open(DEPS, 'r') as file:
+      deps_content = file.read()
+
+    # Eval the content.
+    exec(deps_content, global_scope, local_scope)
+
+    # Extract the deps and filter.
+    deps = local_scope.get('vars')
+
+    if UPSTREAM_PREFIX + dep_name in deps:
+      try:
+        # get the upstream URL from the mapping in DEPS file
+        upstream = deps.get(UPSTREAM_PREFIX + dep_name)
+        # clone dependency from mirror
+        # print(f'attempting: git clone --quiet {dep_name}')
+        # os.system(f'git clone {dep[0]} --quiet {dep_name}')
+        os.system(
+            'git clone {depUrl} --quiet {dep_name}'.format(
```

Comment:
Since I'm not sure how far back in the history the common commit will be, I don't think I could easily do a shallow clone with --depth. But since I'm only interested in the default branch, I could clone just a single branch.